src/fields.cpp

```diff
@@ -472,15 +472,28 @@ bool fields_chunk::alloc_f(component c) {
   return changed;
 }
 
-void fields::require_component(component c) {
-  if (!gv.has_field(c))
-    abort("cannot require a %s component in a %s grid", component_name(c), dimension_name(gv.dim));
-
-  if (beta != 0 && gv.dim != D2) abort("Nonzero beta unsupported in dimensions other than 2.");
+// allocate fields for components required by any source on any process
+// ... this is needed after calling the low-level fields::add_srcdata
+void fields::require_source_components() {
+  int needed[NUM_FIELD_COMPONENTS];
+  memset(needed, 0, sizeof(needed));
+  for (int i = 0; i < num_chunks; i++) {
+    FOR_FIELD_TYPES(ft) {
+      for (src_vol *src = chunks[i]->sources[ft]; src; src = src->next)
+        needed[src->c] = 1;
+    }
+  }
+  int allneeded[NUM_FIELD_COMPONENTS];
+  or_to_all(needed, allneeded, NUM_FIELD_COMPONENTS);
```

Comment:
For accurate timing statistics, it would be useful to wrap this `or_to_all` statement with `am_now_working_on(MpiAllTime);` and `finished_working();`.

Reply:
Done.